proposals/context.md

```diff
@@ -0,0 +1,265 @@
+
+# Goals
+
+- Allow elements in the DOM to retrieve data based on their contextual position in the DOM
+- Alleviate the problem of 'prop drilling'
+- Simple API that is easily implemented in any framework / library
+- Synchronous protocol, while supporting asynchronous patterns
+- Support single or multiple delivery of context values
+
+# Non-Goals
+
+**Context API !== Dependency Injection Framework**
+
+The Context API does not intend to cover all cases and forms of Dependency Injection. It does not specify constructor, factory or property injection patterns. Its only goal is to formalize the pattern of sharing data across the hierarchy in the DOM, specifically avoiding 'prop drilling' type scenarios. Dependency Injection patterns could be implemented using this protocol, but this is not the goal and should remain explicitly outside the scope of Context API for simplicity.
+
+**Context API is not a state management alternative**
+
+State management libraries often need to perform similar behaviors to the problems that Context API helps to solve. An element deep in the DOM tree made need access to some state, and may need to respond to that state being changed. While state management could be built using the Context API, it is not a primary goal of Context API to solve this problem.
+
+# Details
+
+## The `request-context` event
+
+Components which wish to receive some data from their ancestors should initiate the request by firing a composed, bubbling, `request-context` Event, with a `callback` property.
+
+Typescript interface:
+
+```typescript
+interface RequestContextEvent extends Event {
+  /**
+   * The name of the context that is requested
+   */
+  readonly name: T;
+  /**
+   * A boolean indicating if the context should only be provided once.
+   */
+  readonly once: boolean;
+  /**
+   * A callback which a provider of this named callback should invoke.
+   */
+  readonly callback: RequestContextCallback<ContextTypeMap[T]>;
+}
+```
+
+A full typescript definition for this event and its associated types can be found in the [Definitions](#definitions) section at the end of this document.
+
+## Context Providers
+
+A context provider will satisfy a `request-context` event, passing the `callback` the requested data whenever the data changes. A provider will attach an event listener to the DOM tree to catch the event, and if it will be able to satisfy the request will call `stopPropagation` on the event.
+
+If the provider has data available to satisfy the request then it should immediately invoke the `callback` passing the data. If the event does NOT have a truthy `once` property, then the provider can assume that the `callback` can be invokved multiple times, and may retain a reference to the callback to invoke as the data changes. If this is the case the provider should pass the second `dispose` parameter to the callback when invoking it in order to allow the requester to disconnect itself from the providers notifications.
```

Comment:
```suggestion
If the provider has data available to satisfy the request then it should immediately invoke the `callback` passing the data. If the event does NOT have a truthy `once` property, then the provider can assume that the `callback` can be invokved multiple times, and may retain a reference to the callback to invoke as the data changes. If this is the case the provider must pass the second `dispose` parameter to the callback when invoking it in order to allow the requester to disconnect itself from the providers notifications.
```

Add "must"? I think the key requirement is that if a provider may call the callback multiple times, it must provide a disconnect function, so that the converse is true: a defined disconnect function is an indicator of whether or not the callback can be called multiple times.